source/common/config/watch_map.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <set>
+#include <string>
+#include <utility>
+
+#include "envoy/config/subscription.h"
+
+#include "common/common/assert.h"
+#include "common/common/logger.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+
+namespace Envoy {
+namespace Config {
+
+// Manages "watches" of xDS resources. Several xDS callers might ask for a subscription to the same
+// resource name "X". The xDS machinery must return to each their very own subscription to X.
+// The xDS machinery's "watch" concept accomplishes that, while avoiding parallel redundant xDS
+// requests for X. Each of those subscriptions is viewed as a "watch" on X, while behind the scenes
+// there is just a single real subscription to that resource name.
+//
+// This class maintains the watches<-->subscription mapping: it
+// 1) delivers updates to all interested watches, and
+// 2) tracks which resource names should be {added to,removed from} the subscription when the
+//    {first,last} watch on a resource name is {added,removed}.
+//
+// #1 is accomplished by WatchMap's implementation of the SubscriptionCallbacks interface.
+// This interface allows the xDS client to just throw each xDS update message it receives directly
+// into WatchMap::onConfigUpdate, rather than having to track the various watches' callbacks.
+//
+// The information for #2 is returned by updateWatchInterest(); the caller should use it to
+// update the subscription accordingly.
+//
+// A WatchMap is assumed to be dedicated to a single type_url type of resource (EDS, CDS, etc).
+class WatchMap : public SubscriptionCallbacks, public Logger::Loggable<Logger::Id::config> {
+public:
+  WatchMap() {}
+
+  // An opaque token given out to users of WatchMap, to identify a given watch.
+  using Token = uint64_t;
+
+  // Adds 'callbacks' to the WatchMap, with no resource names being watched.
+  // (Use updateWatchInterest() to add some names).
+  // Returns a new token identifying the newly added watch.
+  Token addWatch(SubscriptionCallbacks& callbacks);
+
+  // Returns true if this was the very last watch in the map.
+  // Expects that the watch to be removed has already had all of its resource names removed via
+  // updateWatchInterest().
+  bool removeWatch(Token token);
+
+  // Updates the set of resource names that the given watch should watch.
+  // Returns any resource name additions/removals that are unique across all watches. That is:
+  // 1) if 'resources' contains X and no other watch cares about X, X will be in pair.first.
+  // 2) if 'resources' does not contain Y, and this watch was the only one that cared about Y,
```

Comment:
Why is this pair interesting? It seems a pretty complicated set of relationships, ideally it would be nicer if each watcher only needs to concern themselves with their own resources and no other watcher's.

Reply:
Good question. This returned pair is meant to be used by the multiplexer, not the watches. The context of how I plan to use this class should help clarify what's going on: https://github.com/fredlas/envoy/blob/ADS_incremental/source/common/config/grpc_delta_xds_context.h

Reply:
It seems like a pretty complicated cross layering abstraction breach :) Maybe `GrpcDeltaXdsContext` can register callbacks to learn when resources are interesting?

Reply:
I have to disagree with that; I see this setup as keeping things very cleanly separated. The Context/mux's job is to mediate between the Subscription objects Envoy uses, and the gRPC streams carrying the subscription protocol to the server. Furthermore, there is this watch concept, where multiple Subscription objects can overlap on the same resource - which is a complication the mux would really prefer didn't exist. WatchMap allows the mux to offload all of the watch logic. The mux simply tells it "here's what's going on with your watch stuff" and gets "here's what's going on with your subscription stuff" in response. Compare with GrpcMuxImpl, where the watch and subscription concepts are so subtly interwoven that I actually didn't realize that watches were more than an internal implementation choice until a couple of weeks ago.

WatchMap isn't doing anything about tracking the subscriptions, but it is natural for it to understand the concept of "do I have any watches that care about this resource", which is all the context/mux needs to know to manage its subscriptions.

The subscription interest only ever changes at the same time watch interest changes, so getting subscription changes returned from applying the watch changes flows naturally. I think callbacks would make it more complicated.

Reply:
I think we've gone to one complicated interweaved implementation to two easier to reason about implementations that still have a complex interdependency. Arguably this is an improvement, in that it makes things explicit, but I'd be interested in any way to make these less dependent on each other.

Maybe one way to approach this is to provide a separate set of methods for the WatchMap to tell how many watchers care about a resource. This decouples it from the verb of `updateWatchInterest`.

Reply:
Which parts do you consider interdependency? Here are the interactions I see:

* SubscriptionState knows nothing about watches.
* The multiplexor knows basically nothing about watches *or* subscriptions, other than to pass watch-related stuff through to WatchMap, and subscription-related stuff through to SubscriptionState. My current code over in ADS_incremental would need a little cleanup to show this, but mux's code to handle a watch update can literally be `sub_state.updateSubscriptionInterest(watch_map.updateWatchInterest(resources))`.
* WatchMap tracks nothing about subscription, and arguably "knows" nothing about it. It has a function that tells it to gain/lose watches on resource names, which returns the list of those gained/lost names that are the first/last of their kind. That seems like a reasonable thing for an outsider to want to know when adding/removing things.

I don't see how this could be any more cleanly separated, given that the list of resources subscribed is dictated by the list of resources watched. There has to be translation at some point. I'll attach an illustration I've been meaning to eventually include with this code, updated to include watches.

> Maybe one way to approach this is to provide a separate set of methods for the WatchMap to tell how many watchers care about a resource. This decouples it from the verb of updateWatchInterest.

Why split something into two functions that need to be called one after another? That opens the door to forgetting to call both. More importantly, I think what you're suggesting is impossible. If you've already updated the names, you no longer have the prior state to diff against, meaning you no longer have a way to know what names were removed.

Reply:
![delta_xDS_code_flow](https://user-images.githubusercontent.com/43351173/58995909-09d7fb00-87c4-11e9-867b-3b41f42f87b4.png)

Reply:
I've had some time to reflect on this, and I think the main issue that I found confusing when looking at this design is the implicit contract that exists, where `WatchMap` is in the business of both (1) demuxing/dispatching updates _and_ (2) also in supporting a client that is trying to maintain a set of subscriptions of relevance to the watches. It feels like these are almost two different modules. This is due to the blue arrow above the "pass through" `GrpcXdsContext`.

What would it look like if `GrpcXdsContext` did its own lightweight tracking for (2)? It would be technically redundant and wasteful of memory, but it might be conceptually cleaner. WDYT?

Reply:
(sorry, I didn't see that this had gone into this thread when replying; I only saw the unreply-able one.)

My first reaction was that this made sense, but when I looked closer, it doesn't work. They may seem like two different things, but really, there is just a single thing that WatchMap does: it tracks which watches are interested in which resource names. (1) is something that needs what WatchMap tracks, so it makes sense for it to do it. (2) is not actually something that needs to be tracked, but rather, is a side-effect of making changes to that tracking. I don't see how anything could possibly be cleaner than expressing that as "here are some names to add/remove to your watch map, btw are any of them the first/last references to those names?"

Have you looked at the GrpcDeltaXdsContext in the other branch I linked to? I think looking at how it uses WatchMap should help you understand how extremely cleanly this logic is all separated out.

Reply:
Yeah, I get how it simplifies `GrpcDeltaXdsContext`, the main downside is the interface is really very specific to it. I think it probably make sense to disagree and commit at this point, as I think `WatchMap` can work reasonably, it's just not the cleanest abstraction from my POV (there's no accounting for taste..). So, let's go with what you have.

I'll review the implementation, meanwhile, could you put that diagram in some docs inside the PR as well? It's quite helpful.

Reply:
Ok, thanks. One thing that might make the introduction of this thing more appealing: this whole framework centered on `GrpcDeltaXdsContext` (name not final) is going to be able to entirely provide SotW xDS/ADS with a trivial amount of additional work.

Will add shortly.

Reply:
Added diagram.